database/dbService.js

```diff
@@ -0,0 +1,84 @@
+var mysql = require('mysql');
+var keys = require('../config/keys');
+var queries = require('./queries');
+
+var connection;
+module.exports = {
+    /* Inserts a given user model given by Google OAuth */
+    insertUser: function(user, callback) {
+        this.tryConnect().getConnection(function(err, con) {
+            var sql = queries.insertUser;
+            con.query(sql, [user.id, user.gender, user.name.givenName, user.name.familyName, user.emails[0].value], function (err, result) {
+                con.release();
+                if (err) throw err;
+                callback(result.insertId);
+            });
+        });
+    },
+
+    /* Retrieves a User model by GoogleID */
+    getUserByGoogleID: function(googleID, callback) {
+        this.tryConnect().getConnection(function(err, con) {
+            var sql = queries.getUserByGoogleID;
+            con.query(sql, googleID, function (err, result) {
+                con.release();
+                if (err) throw err;
+                // Call the callback function in the caller of this method so we can do something with this "result"
+                return callback(result); // [] if not found
+            });
+        });
+    },
+
+    /* Retrieves a User model by ID */
+    getUserByID: function(ID, callback) {
+        this.tryConnect().getConnection(function(err, con) {
+            var sql = queries.getUserByID;
+            con.query(sql, ID, function (err, result) {
+                con.release();
+                if (err) throw err;
+                // Call the callback function in the caller of this method so we can do something with this "result"
+                return callback(result); // [] if not found
+            });
+        });
+    },
+
+    tryConnect: function() {
+        // Singleton: if connection is already established, return it
+        if (connection) {
+            console.log("Getting connection...already made");
+            return connection;
+        }
+
+        // Otherwise, recreate the connection since the old one cannot be reused (due to either errors or upon initial app start-up)
+        connection = mysql.createPool({
+            connectionLimit: 100,
```

Comment:
I'm using Pools, which means it's a pool of connections. 100 connections max.

SOMETHING I WILL PUT IN THIS PULL REQUEST (NOT DONE YET):
Every time we make a query with connection, right now we return the existing connection (singleton). This is not optimized. This means we lock on to one connection, run the query, release the lock and continue with the next query with the same connection.

That's sequential. We need it to be async. Let's use a new connection for every query. Thus, the refactor means I have to remove the singleton logic.

Reply:
@trevorhalvorson @Kgrah for visibility.

Reply:
This is now done.